File: src/backend/IAW.Vdf.Persistence/Seeding/RulesCorpusImporter.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using IAW.Vdf.Core.Serialization;
using IAW.Vdf.Persistence.Entities;
using IAW.Vdf.Persistence.Repositories;
using Microsoft.EntityFrameworkCore;

namespace IAW.Vdf.Persistence.Seeding;

/// <summary>
/// Imports the rules corpus (a directory of <c>*.json</c> rule files) and the
/// <c>reference-data.json</c> file into the Postgres database via <see cref="EfRuleRepository"/>
/// and the <see cref="VdfDbContext"/> directly for reference data.
///
/// Rules are upserted using <see cref="EfRuleRepository.SaveAsync"/> so versioning semantics are
/// honoured. Reference data is upserted by (Source, Key) pairs, flattening the JSON object
/// hierarchy into individual rows.
/// </summary>
public sealed class RulesCorpusImporter
{
    private readonly EfRuleRepository _ruleRepo;
    private readonly VdfDbContext _db;

    /// <summary>Creates the importer.</summary>
    /// <param name="ruleRepo">The EF rule repository.</param>
    /// <param name="db">The VDF database context.</param>
    public RulesCorpusImporter(EfRuleRepository ruleRepo, VdfDbContext db)
    {
        _ruleRepo = ruleRepo;
        _db = db;
    }

    /// <summary>
    /// Imports all rules from <paramref name="rulesDirectory"/> and reference data from
    /// <paramref name="referenceDataPath"/> into Postgres.
    /// </summary>
    /// <param name="rulesDirectory">Path to the directory containing <c>*.json</c> rule files.</param>
    /// <param name="referenceDataPath">Path to <c>reference-data.json</c>.</param>
    /// <param name="cancellationToken">A cancellation token.</param>
    public async Task ImportAsync(
        string rulesDirectory,
        string referenceDataPath,
        CancellationToken cancellationToken = default)
    {
        await ImportRulesAsync(rulesDirectory, cancellationToken).ConfigureAwait(false);
        await ImportReferenceDataAsync(referenceDataPath, cancellationToken).ConfigureAwait(false);
    }

    private async Task ImportRulesAsync(string rulesDirectory, CancellationToken cancellationToken)
    {
        var files = Directory.EnumerateFiles(rulesDirectory, "*.json", SearchOption.TopDirectoryOnly)
            .Where(f => !Path.GetFileName(f).Equals("reference-data.json", StringComparison.OrdinalIgnoreCase))
            .OrderBy(f => f, StringComparer.Ordinal);

        foreach (var file in files)
        {
            cancellationToken.ThrowIfCancellationRequested();
            var json = await File.ReadAllTextAsync(file, cancellationToken).ConfigureAwait(false);
            var trimmed = json.TrimStart();

            if (trimmed.StartsWith('['))
            {
                var rules = RuleSerializer.DeserializeMany(json);
                foreach (var rule in rules)
                {
                    await _ruleRepo.SaveAsync(rule, cancellationToken).ConfigureAwait(false);
                }
            }
            else
            {
                var rule = RuleSerializer.Deserialize(json);
                await _ruleRepo.SaveAsync(rule, cancellationToken).ConfigureAwait(false);
            }
        }
    }

    private async Task ImportReferenceDataAsync(string referenceDataPath, CancellationToken cancellationToken)
    {
        if (!File.Exists(referenceDataPath))
        {
            return;
        }

        var json = await File.ReadAllTextAsync(referenceDataPath, cancellationToken).ConfigureAwait(false);
        var root = JsonNode.Parse(json) as JsonObject
            ?? throw new InvalidOperationException("reference-data.json must be a JSON object.");

        foreach (var (topKey, topValue) in root)
        {
            if (topValue is null) continue;

            if (topValue is JsonObject nestedObj)
            {
                // Flatten: Source=topKey, Key=each nested property
                foreach (var (nestedKey, nestedValue) in nestedObj)
                {
                    await UpsertReferenceDataAsync(topKey, nestedKey, nestedValue?.ToJsonString() ?? "null", cancellationToken)
                        .ConfigureAwait(false);
                }
            }
            else
            {
                // Scalar or array at top level: Source=topKey, Key=""
                await UpsertReferenceDataAsync(topKey, "", topValue.ToJsonString(), cancellationToken)
                    .ConfigureAwait(false);
            }
        }
    }

    private async Task UpsertReferenceDataAsync(
        string source, string key, string valueJson, CancellationToken cancellationToken)
    {
        var existing = await _db.ReferenceData
            .FirstOrDefaultAsync(r => r.Source == source && r.Key == key, cancellationToken)
            .ConfigureAwait(false);

        if (existing is null)
        {
            _db.ReferenceData.Add(new ReferenceDataEntity
            {
                Id = Guid.NewGuid(),
                Source = source,
                Key = key,
                ValueJson = valueJson,
            });
        }
        else
        {
            existing.ValueJson = valueJson;
        }

        await _db.SaveChangesAsync(cancellationToken).ConfigureAwait(false);
    }
}

File: src/server/src/rules/rules-corpus.importer.ts
import { Injectable, Logger } from '@nestjs/common';
import * as fs from 'fs';
import * as path from 'path';
import { Prisma } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { deserializeRule } from '../vdf/serializer';
import { JsonObject, JsonValue } from '../vdf/types';
import { RuleRepository } from './rule.repository';

const SYSTEM_ACTOR = 'system';

/** The number of rules and reference-data entries imported. */
export interface ImportResult {
  rulesImported: number;
  referenceEntriesImported: number;
}

/**
 * Idempotent importer for the on-disk corpus.
 *
 * Loads every `rules/*.json` (excluding `reference-data.json`) into Rule+RuleVersion
 * via {@link RuleRepository.saveAsync} (so v1 is created with full versioning
 * semantics), and flattens `reference-data.json` into `reference_data` rows mirroring
 * the .NET RulesCorpusImporter:
 *  - nested object source → one row per nested key (key=nestedKey),
 *  - scalar/array source (incl. literal dotted keys) → one row with key="".
 *
 * Reference data is upserted by (source, key) so repeated imports are safe. Rules are
 * imported only when the Rule table is empty ({@link ensureImported}), or
 * unconditionally via {@link importCorpus} (each call appends a new version).
 */
@Injectable()
export class RulesCorpusImporter {
  private readonly logger = new Logger(RulesCorpusImporter.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly ruleRepo: RuleRepository,
  ) {}

  /** Imports the corpus iff no rules exist yet. Returns counts (zeros when skipped). */
  async ensureImported(rulesDir?: string): Promise<ImportResult> {
    const existing = await this.prisma.rule.count();
    if (existing > 0) {
      this.logger.log(
        `Corpus already imported (${existing} rules) — skipping.`,
      );
      return { rulesImported: 0, referenceEntriesImported: 0 };
    }
    return this.importCorpus(rulesDir);
  }

  /**
   * Imports all rule files and the reference-data file from {@link rulesDir}
   * (defaults to the repo-root `rules/` located by walking up from this file).
   */
  async importCorpus(rulesDir?: string): Promise<ImportResult> {
    const dir = rulesDir ?? findRulesDir();
    const rulesImported = await this.importRules(dir);
    const referenceEntriesImported = await this.importReferenceData(
      path.join(dir, 'reference-data.json'),
    );
    this.logger.log(
      `Imported ${rulesImported} rules and ${referenceEntriesImported} reference entries.`,
    );
    return { rulesImported, referenceEntriesImported };
  }

  private async importRules(dir: string): Promise<number> {
    const files = fs
      .readdirSync(dir)
      .filter((f) => f.endsWith('.json') && f !== 'reference-data.json')
      .sort();

    let count = 0;
    for (const file of files) {
      const json = fs.readFileSync(path.join(dir, file), 'utf8');
      const trimmed = json.trimStart();
      if (trimmed.startsWith('[')) {
        const arr = JSON.parse(json) as unknown[];
        for (const node of arr) {
          const rule = deserializeRule(JSON.stringify(node));
          await this.ruleRepo.saveAsync(rule, { authoredBy: SYSTEM_ACTOR });
          count += 1;
        }
      } else {
        const rule = deserializeRule(json);
        await this.ruleRepo.saveAsync(rule, { authoredBy: SYSTEM_ACTOR });
        count += 1;
      }
    }
    return count;
  }

  private async importReferenceData(refPath: string): Promise<number> {
    if (!fs.existsSync(refPath)) {
      return 0;
    }
    const parsed = JSON.parse(fs.readFileSync(refPath, 'utf8')) as JsonObject;

    let count = 0;
    for (const [topKey, topValue] of Object.entries(parsed)) {
      if (topValue === null || topValue === undefined) {
        continue;
      }
      if (
        typeof topValue === 'object' &&
        !Array.isArray(topValue) &&
        // A literal dotted top-level key (e.g. "TestCompendium.compatibleSpecimens")
        // is stored whole; only PLAIN nested objects fan out. Literal dotted keys are
        // arrays/objects keyed by a dotted name and are handled by the else branch.
        !topKey.includes('.')
      ) {
        for (const [nestedKey, nestedValue] of Object.entries(topValue)) {
          await this.upsert(topKey, nestedKey, nestedValue ?? null);
          count += 1;
        }
      } else {
        await this.upsert(topKey, '', topValue);
        count += 1;
      }
    }
    return count;
  }

  private async upsert(
    source: string,
    key: string,
    value: JsonValue,
  ): Promise<void> {
    const valueJson = (
      value === undefined ? null : value
    ) as Prisma.InputJsonValue;
    await this.prisma.referenceDataEntry.upsert({
      where: { source_key: { source, key } },
      create: { source, key, valueJson },
      update: { valueJson },
    });
  }
}

/**
 * Walks up from this file to find the repo-root corpus `rules/` directory — the one
 * containing `reference-data.json`. The marker disambiguates it from the server's own
 * `src/rules/` source module, which would otherwise shadow the walk-up.
 */
function findRulesDir(): string {
  let dir: string | undefined = __dirname;
  while (dir !== undefined) {
    const candidate = path.join(dir, 'rules');
    if (
      fs.existsSync(candidate) &&
      fs.statSync(candidate).isDirectory() &&
      fs.existsSync(path.join(candidate, 'reference-data.json'))
    ) {
      return candidate;
    }
    const parent = path.dirname(dir);
    dir = parent === dir ? undefined : parent;
  }
  throw new Error(
    `Could not locate corpus 'rules' directory from ${__dirname}.`,
  );
}
